Fix markdown types.

showell opened the pull request.

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1876,10 +1876,12 @@ def handleMatch(  # type: ignore[override] # supertype incompatible with supersu
                     self.md.zulip_rendering_result.mentions_wildcard = True
                 user_id = "*"
             elif user:
+                assert type(user) == FullNameInfo
```
This should be an `isinstance` check.

showell replied:
I don't think it really matters since we're never gonna subclass `FullNameInfo`.

andersk replied:
The right idiom when it doesn’t matter is `isinstance`, according to PEP 8.

showell replied:
One curiosity here is that I need to add the type check here in order for mypy to complain that I had had `user["full_name"]` instead of `user.full_name`.  If you convert the `if user` check to `if user is not None` then mypy still can't apparently infer that `user` is a type of `FullNameInfo`.  Fortunately, the test suite quickly alerted me to the code that I needed to change.

Anyway, I am converting the assertion to use `isinstance`.  (Either flavor of assertion helps mypy produce an error along the lines of "Value of type "FullNameInfo" is not indexable" if you fail to cut over to attribute access.)

showell replied:
(I also changed the check from `elif user` to `ellif user is not None`.)

timabbott commented on zerver/lib/mention.py:
```diff
@@ -21,6 +21,22 @@ class FullNameInfo:
+        elif self.full_name is not None:
+            return Q(full_name__iexact=self.full_name)
+        else:
+            raise AssertionError("totally empty filter makes no sense")
```
I don't understand where you're going with this extraction; it seems like having the `Q` object return an `Any` isn't ideal.

showell replied:
I can explain this better over chat, but the idea behind this extraction is to allow us to efficiently send a bunch of PMs to a bunch of users when we subscribe a bunch of users to a stream.

It sucks that the Q() function returns an Any type, but mypy won't accept `object`, and Django and mypy generally don't play nice.  (I can explain that in person or over chat.)

showell replied:
In the next day or two I will push the rest of the PR, and this decision will make more sense.  In the meantime, the first few commits of this PR are probably no-brainers.

andersk replied:
The right annotation is `def Q(self) -> Q`, since the value has type `django.db.models.Q` at runtime.

For now, mypy will treat that as equivalent to `-> Any` because it doesn’t have stubs for `django.db.models.Q`, but it will be its own type once we add django-stubs (#19552).

showell replied:
> The right annotation is `def Q(self) -> Q`, since the value has type `django.db.models.Q` at runtime.
> 
> For now, mypy will treat that as equivalent to `-> Any` because it doesn’t have stubs for `django.db.models.Q`, but it will be its own type once we add django-stubs (#18777).

Arrrrrrrrrrrrrrrrrg! I knew that Q was probably Any at heart but it somehow didn’t occur to me that the type of Q is Q.  Omg

showell replied:
Who is buried in Grant’s Tomb?

showell replied:
I always wondered what was happening under the hood.  See `/django/db/models/query_utils.py`.

Ok, next push will use Q as the return type.  (For some reason I was thinking Q was a function returning some other data type.)

~~~ py
 57 class Q(tree.Node):
 58     """
 59     Encapsulate filters as objects that can then be combined logically (using
 60     `&` and `|`).
 61     """
 62     # Connection types
 63     AND = 'AND'
 64     OR = 'OR'
 65     default = AND
 66     conditional = True
 67
 68     def __init__(self, *args, _connector=None, _negated=False, **kwargs):
 69         super().__init__(children=[*args, *sorted(kwargs.items())], connector=_connector, negated=_negated)
 70
[...]

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -134,7 +134,16 @@ class MessageRenderingResult:
+    mention_data: MentionData
+    realm_uri: str
+    realm_alert_words_automaton: Optional[ahocorasick.Automaton]
+    active_realm_emoji: Dict[str, Dict[str, str]]
```
This annotation is wrong; it looks like this was propagated from an existing wrong annotation. An emoji dict seems to have `id: str, name: str, source_url: str, deactivated: bool, author_id: int`.

showell replied:
I added a commit to address this.  It's slightly annoying due to `still_url`, which I explain in the commit message.

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1506,7 +1515,7 @@ def handleMatch(self, match: Match[str]) -> Optional[Union[str, Element]]:
         active_realm_emoji: Dict[str, Dict[str, str]] = {}
```
(source of the existing wrong annotation)

andersk replied:
If this incorrect annotation still type checks, then something else is wrong, since it shouldn’t be compatible with the corrected type of `db_data.active_realm_emoji`.

showell replied:
This is due to the fact that the `markdown` library doesn't have a good story for plumbing though context data like our `DbData` stuff, so we have to attach the db data in a way that confuses mypy.  I am attempting something here to at least mitigate that.  Will report back.

showell replied:
~~~
$ git diff
diff --git a/zerver/lib/markdown/__init__.py b/zerver/lib/markdown/__init__.py
index 00b4369142..4c4a3f263f 100644
--- a/zerver/lib/markdown/__init__.py
+++ b/zerver/lib/markdown/__init__.py
@@ -995,7 +995,7 @@ class InlineInterestingLinkProcessor(markdown.treeprocessors.Treeprocessor):
             else:
                 current_node.tail = text

-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data
         current_index = 0
         for item in to_process:
             # The text we want to link starts in already linked text skip it
@@ -1330,7 +1330,7 @@ class InlineInterestingLinkProcessor(markdown.treeprocessors.Treeprocessor):
                 # is enabled, but URL previews are a beta feature and YouTube
                 # previews are pretty stable.

-            db_data = self.md.zulip_db_data
+            db_data: Optional[DbData] = self.md.zulip_db_data
             if db_data and db_data.sent_by_bot:
                 continue

@@ -1486,7 +1486,7 @@ class EmoticonTranslation(markdown.inlinepatterns.Pattern):
     """Translates emoticons like `:)` into emoji like `:smile:`."""

     def handleMatch(self, match: Match[str]) -> Optional[Element]:
-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data
         if db_data is None or not db_data.translate_emoticons:
             return None

@@ -1513,7 +1513,7 @@ class Emoji(markdown.inlinepatterns.Pattern):
         name = orig_syntax[1:-1]

         active_realm_emoji: Dict[str, Dict[str, str]] = {}
-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data
         if db_data is not None:
             active_realm_emoji = db_data.active_realm_emoji

@@ -1625,7 +1625,7 @@ class CompiledPattern(markdown.inlinepatterns.Pattern):
 class AutoLink(CompiledPattern):
     def handleMatch(self, match: Match[str]) -> ElementStringNone:
         url = match.group("url")
-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data
         return url_to_a(db_data, url)


@@ -1827,7 +1827,7 @@ class LinkifierPattern(CompiledInlineProcessor):
     def handleMatch(  # type: ignore[override] # supertype incompatible with supersupertype
         self, m: Match[str], data: str
     ) -> Union[Tuple[Element, int, int], Tuple[None, None, None]]:
-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data
         url = url_to_a(
             db_data,
             self.format_string % m.groupdict(),
@@ -1849,7 +1849,7 @@ class UserMentionPattern(CompiledInlineProcessor):
     ) -> Union[Tuple[None, None, None], Tuple[Element, int, int]]:
         name = m.group("match")
         silent = m.group("silent") == "_"
-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data
         if db_data is not None:
             wildcard = mention.user_mention_matches_wildcard(name)

@@ -1903,7 +1903,7 @@ class UserGroupMentionPattern(CompiledInlineProcessor):
     ) -> Union[Tuple[None, None, None], Tuple[Element, int, int]]:
         name = m.group("match")
         silent = m.group("silent") == "_"
-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data

         if db_data is not None:
             user_group = db_data.mention_data.get_user_group(name)
@@ -1932,7 +1932,7 @@ class UserGroupMentionPattern(CompiledInlineProcessor):

 class StreamPattern(CompiledInlineProcessor):
     def find_stream_id(self, name: str) -> Optional[int]:
-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data
         if db_data is None:
             return None
         stream_id = db_data.stream_names.get(name)
@@ -1963,7 +1963,7 @@ class StreamPattern(CompiledInlineProcessor):

 class StreamTopicPattern(CompiledInlineProcessor):
     def find_stream_id(self, name: str) -> Optional[int]:
-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data
         if db_data is None:
             return None
         stream_id = db_data.stream_names.get(name)
@@ -2028,7 +2028,7 @@ class AlertWordNotificationProcessor(markdown.preprocessors.Preprocessor):
         return False

     def run(self, lines: List[str]) -> List[str]:
-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data
         if db_data is not None:
             # We check for alert words here, the set of which are
             # dependent on which users may see this message.
@@ -2062,7 +2062,7 @@ class LinkInlineProcessor(markdown.inlinepatterns.LinkInlineProcessor):
             return None  # no-op; the link is not processed.

         # Rewrite local links to be relative
-        db_data = self.md.zulip_db_data
+        db_data: Optional[DbData] = self.md.zulip_db_data
         href = rewrite_local_links_to_relative(db_data, href)

         # Make changes to <a> tag attributes
(zulip-py3-venv) ~/zulip (markdown_data|REBASE-i 5/9)
$ ./tools/run-mypy
zerver/lib/markdown/__init__.py:1518: error: Incompatible types in assignment (expression has type "Dict[str, EmojiInfo]", variable has type "Dict[str, Dict[str, str]]")  [assignment]
Found 1 error in 1 file (checked 1326 source files)

showell replied:
See the commit entitled "mypy: Add explicit types for dbdata references. " that tries to plug that loophole for now.  (It's obviously not completely robust for the long term.)

And then the specific mypy issue is now addressed as part of the EmojiInfo commit.  I put the EmojiInfo commit early in the sequence, but it's not till the subsequent commits that mypy has any teeth related to the type errors.

showell replied:
(ready for review again)

showell commented:
Ok, I believe I've addressed all the feedback here.  Please hold off on merging the last commit, as I want to get some better test coverage on it, but I decided to push it, since it puts some of the other commits in more context.

showell commented:
Ok, I improved the test coverage, so this whole PR is ready for review from my perspective.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -13931,8 +13931,9 @@ components:
+          nullable: true
           description: |
-            Only present when the emoji's image is animated.
+            Only non-null when the emoji's image is animated.
```
This requires an API feature level bump if we change it.  (So does basically any other change in this file that isn't copyediting).  It's a necessary habit to allow the mobile apps to be implemented correctly with Zulip productions servers like Zulip Cloud that may run a version between major releases.

I also think we shouldn't make API changes like this just for mypy reasons -- one can use https://mypy.readthedocs.io/en/stable/more_types.html#mixing-required-and-non-required-items to make mypy happy.  But I suspect this is an improvement -- can you start an "#api design" conversation about this proposed API change and mention Greg to get quick feedback?  I suspect that the `nullable` API is better anyway, but it's worth using our API change process for this sort of decision.

andersk replied:
Also Mypy 0.930 now [supports](https://mypy-lang.blogspot.com/2021/12/mypy-0930-released.html) `typing_extensions.NotRequired` for individual `TypedDict` items ([PEP 655](https://www.python.org/dev/peps/pep-0655/)), which is a bit nicer than subclassing with `total=False` just to get some non-required items.

It is true, however, that non-required items weaken Mypy’s checking: Mypy won’t protect you from implicitly assuming the item is present when it might not be. So the API change still has potential value.

showell replied:
I am not a fan of `NotRequired` fields, unless there are extreme performance requirements about what you send over the wire.  As Anders alludes to, it makes it difficult to know whether a codepath left off the value due to negligence or due to a proper business rule.

I just bumped the API version.  I am 99% sure that mobile does not try to use the field currently.  I haven't pushed yet, but this is the gist of my edit:

~~~
13946             **Changes**: New in Zulip 5.0 (added as optional field in
13947             feature level 97 and then made mandatory, but nullable, in
13948             feature level 113).

timabbott commented on zerver/lib/mention.py:
```diff
@@ -37,27 +37,58 @@ def Q(self) -> Q:
+                    user = self.cache.get((user_filter.id, user_filter.full_name), None)
+
+            if user is not None:
+                result.append(user)
```
Could this be written more readably with `if self.cache:` and then doing e.g. `assert user_filter.id is not None`, moving this if statement inside the `self.cache` block, and doing a `continue` after the `result.append` line rather than an else clause?  I was having trouble thinking about all the fallthrough situations in this function.

timabbott replied:
Hmm I guess the assert part doesn't work since callers aren't explicit about whether they want the cache, but the rest of the ideas seem helpful.

showell replied:
I don't understand the `continue` suggestion.  The idea in the first loop is that you either get a cache hit (and append to results) or a cache miss (and then need to add to the list of things to do in the second loop).  I tried tweaking a few variables and comments here, so hopefully on the next review the code will make a bit more sense.

andersk replied:
I believe the suggestion is

```diff
 for user_filter in user_filters:
-    user: Optional[FullNameInfo] = None
-
     if user_filter.id is not None and user_filter.full_name is not None:
         user = self.user_cache.get((user_filter.id, user_filter.full_name), None)
+        if user is not None:
+            result.append(user)
+            continue
 
-    if user is not None:
-        result.append(user)
-    else:
-        unseen_user_filters.append(user_filter)
+    unseen_user_filters.append(user_filter)
```

showell replied:
Ah, ok, fix coming.

showell replied:
re-pushed

timabbott commented:
This looks great other than the two issues noted above.

showell commented:
I believe I have addressed the comments, so this is ready for review again.

showell commented:
I also added a couple commits related to stream mentions.

timabbott commented:
Merged, after a small tweak to the API changelog entry to match our standard format of listing all the affected endpoints.

(I just copied the list from the feature level 97 entry).  Thanks @showell!
